### ISSUE TITLE
Simplify plugin settings classes

### ISSUE THREAD

**Issue by Osiris-Team** [fvarrui/JavaPackager#251]
As far as I know, there are 4 classes that are relevant to the plugin settings:
- `PackagePluginExtension extends PackagerSettings`
- `Packager extends PackagerSettings`
- `PackageTask extends AbstractPackageTask` does not extend `PackagerSettings` but has the same fields and methods, the only difference is, that it has additional gradle-specific annotations for its fields and methods `manifest(...) scripts(...) winConfig() linuxConfig() macConfig()` do gradle-specific stuff. It also has a method to create a `Packager` with defaults taken from the `PackagePluginExtension` class.
- `PackageMojo extends AbstractMojo` does not extend `PackagerSettings` but has the same fields and methods, the only difference is, that it has additional maven-specific annotations for its fields. Like the `PackageTask` for gradle it has the ability to create a `Packager`, but the defaults are taken from its own fields.

To simplify this I am proposing to:
- take the annotations of `PackageMojo` and `PackageTask` and put them inside the `PackagerSettings` class. The problem with this is that then the `PackagerSettings` class would have to be extended by `PackageMojo` and `PackageTask`, this won't work however because they are already extending another class. There must be a way of doing this stuff without annotations, gonna research a bit.
- the defaults could already be set for the fields in `PackagerSettings`

still doing some research, feel free to give input.

**Comment by Osiris-Team** [comment]
This could work?
```java
import org.apache.maven.plugins.annotations.Parameter;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.Optional;

public class MySettings {
    public static boolean IS_GRADLE = false;

    @Parameter(defaultValue = "${project.name}", property = "name", required = false)
    @Input
    @Optional
    public String name = (IS_GRADLE ? "default-gradle-name" : "${project.name}");
}
```

**Comment by fvarrui** [comment]
Ok, I've tried to decouple the plugin core from Gradle and Maven. You can find a class called `Context`. I'd like to remove all duplicated tasks, I mean, doing the same for Gradle and for Maven, e.g. generate Launch4j EXE (trying to keep it simple, less code). I also plan to add new tasks and mojos, so the plugin can offer to developers the possibility to use only parts of the plugin (e.g. create mac app, create windows app, generate dmg, generate exe, generate setup, generate deb, ...). This way developers can use a task for generating a Windows app, then do some process on this generated app, and finally generate a Setup for this app. So, in the future, there will be lots of tasks/mojos, with specific parameters, reusing packagers/artifact generators.

**Comment by fvarrui** [comment]
> This could work?
> 
> ```java
> import org.apache.maven.plugins.annotations.Parameter;
> import org.gradle.api.tasks.Input;
> import org.gradle.api.tasks.Optional;
> 
> public class MySettings {
>     public static boolean IS_GRADLE = false;
> 
>     @Parameter(property = "name", required = false)
>     @Input
>     @Optional
>     public String name = (IS_GRADLE ? "default-gradle-name" : "${project.name}");
> }
> ```
> 
> Still same issue with inheritance though...

I'm not sure. You should test if this works for both building tools: Gradle and Maven, so, if both tools are able to deal with those annotations. Keep them together, in the same plugin, was a headache, but I got it working. I didn't want to maintain two plugins. That's why I refer to the reusable parts as the core of the plugin.

**Comment by Osiris-Team** [comment]
@fvarrui just updated the comment.

**Comment by fvarrui** [comment]
Feel free to enhance the plugin! If you find a way to simplify the code, it would be great!

**Comment by fvarrui** [comment]
> * `PackagePluginExtension extends PackagerSettings`
> * `Packager extends PackagerSettings`
> * `PackageTask extends AbstractPackageTask` does not extend `PackagerSettings` but has the same fields and methods, the only difference is, that it has additional gradle-specific annotations for its fields and methods `manifest(...) scripts(...) winConfig() linuxConfig() macConfig()` do gradle-specific stuff. It also has a method to create a `Packager` with defaults taken from the `PackagePluginExtension` class.
> * `PackageMojo extends AbstractMojo` does not extend `PackagerSettings` but has the same fields and methods, the only difference is, that it has additional maven-specific annotations for its fields. Like the `PackageTask` for gradle it has the ability to create a `Packager`, but the defaults are taken from its own fields.

An UML class diagram would be great 😆

**Comment by Osiris-Team** [comment]
Available in PR #248

**Comment by Osiris-Team** [comment]
Omg, I finally did it!
Both maven and gradle are now working with unified settings!!!! Checkout my branch @fvarrui 
Also please check the things I marked with TODO, wasnt sure about those.
